src/Planner/PlannerJoins.cpp

```diff
@@ -385,6 +600,203 @@ void buildJoinClause(
+                // TODO(antaljanosbenjamin): Analyze the expressions first, so join clauses are not built unnecessarily.
+                const auto with_key_expression = static_cast<size_t>(std::count_if(
+                    result.begin(), result.end(), [](const JoinClause & clause) { return !clause.getLeftKeyNodes().empty(); }));
+
+                if (result.size() > 1 && with_key_expression != 0 && with_key_expression < result.size())
+                {
+                    result.clear();
+                    result.emplace_back();
+                    buildSimpleJoinClause(
+                        left_dag,
+                        right_dag,
+                        joined_dag,
+                        planner_context,
+                        node,
+                        left_table_expressions,
+                        right_table_expressions,
+                        join_node,
+                        result.front());
+                }
+            }
+            else
+            {
+                auto it = arguments.begin();
+                {
+                    auto & child_res = built_clauses.at(it->get());
+                    result.insert(result.end(), std::make_move_iterator(child_res.begin()), std::make_move_iterator(child_res.end()));
+                }
+                it++;
+
+                for (; it != arguments.end(); it++)
+                {
+                    auto & child_res = built_clauses.at(it->get());
+                    result = makeCrossProduct(result, child_res);
+                }
+            }
+
+            built_clauses.emplace(node.get(), std::move(result));
+        }
+        else
+        {
+            nodes_to_process.pop();
+            JoinClauses clauses;
+            clauses.emplace_back();
+            buildSimpleJoinClause(
+                left_dag,
+                right_dag,
+                joined_dag,
+                planner_context,
+                node,
+                left_table_expressions,
+                right_table_expressions,
+                join_node,
+                clauses.front());
+
+            built_clauses.emplace(node.get(), std::move(clauses));
+        }
+    }
+    return std::move(built_clauses.at(join_expression.get()));
+}
+
+std::pair<JoinClauses, bool /*is_inequal_join*/> buildAllJoinClauses(
+    ActionsDAG & left_join_actions,
+    ActionsDAG & right_join_actions,
+    ActionsDAG & post_join_actions,
+    const PlannerContextPtr & planner_context,
+    const QueryTreeNodePtr & join_expression,
+    const TableExpressionSet & join_left_table_expressions,
+    const TableExpressionSet & join_right_table_expressions,
+    const JoinNode & join_node,
+    const FunctionNode & function_node)
+{
+    const auto & join_algorithms = planner_context->getQueryContext()->getSettingsRef()[Setting::join_algorithm];
+    const auto is_hash_join_enabled = TableJoin::isEnabledAlgorithm(join_algorithms, JoinAlgorithm::HASH)
```

Comment:
> How it works in case of auto? 

I haven't checked how does this work exactly with `auto` for the reason that you mentioned: partial merge doesn't support everything, I just opted to disable this for auto in order to ensure we don't run into weird cases because of this. However I expect that it is handled properly in case of `auto` though, I think there is nothing to prevent the users to write a query with `auto` algorithm that works with `hash`, but not with `partial_merge`.